### PR TITLE
PERF: cython optimizations

### PR DESCRIPTION
**Pull request by jbrockmendel**
cython was emitting some warnings about untyped array lookups in `tslibs.conversion`; added types to those

`arr.fill(val)` is a python call, while `arr[:]` is a C call, so replaced all of those

modernized an old non-python for-loop, some other small cleanups along the way.

We're running out of available optimizations/cleanups in these files.  One thing I'd be interested in changing (but I'm not confident enough to do) is removing uses of `double` and `double_t`, which I _think_ are equivalent to `float64_t`.  But I'm not sure if that is architecture or platform dependent or something.

### CODE REVIEW

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -177,7 +177,7 @@ def round_nsint64(values, mode, freq):
 
     # if/elif above should catch all rounding modes defined in enum 'RoundTo':
     # if flow of control arrives here, it is a bug
-    assert False, "round_nsint64 called with an unrecognized rounding mode"
+    raise ValueError("round_nsint64 called with an unrecognized rounding mode")
```
Comment (gfyoung) [inline pandas/_libs/tslibs/timestamps.pyx]: Given that this is a `def ` function, is this error tested in any way?

**Comment by gfyoung** [comment]
> We're running out of available optimizations/cleanups in these files.

@jbrockmendel : Gosh, I wonder whose **"FAULT"** that is 😉

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment.

**Comment by jbrockmendel** [comment]
BTW any idea about the double_t thing?

**Comment by jreback** [comment]
what is the problem with double_t ?
this is just float64

**Comment by jbrockmendel** [comment]
> what is the problem with double_t ?  this is just float64

Yah that's the thing I want to double-check before changing all the places where we currently use `double` or `double_t` to `float64_t`

**Review by jreback — APPROVED** [review]

**Pull request merged**